src/dawn/Optimizer/PassSetBoundaryCondition.cpp

```diff
@@ -0,0 +1,332 @@
+///
+class VisitStencilCalls : public ASTVisitorForwarding {
+  std::vector<std::shared_ptr<StencilCallDeclStmt>> stencilCallsInOrder_;
+
+public:
+  const std::vector<std::shared_ptr<StencilCallDeclStmt>>& getStencilCalls() const {
+    return stencilCallsInOrder_;
+  }
+  std::vector<std::shared_ptr<StencilCallDeclStmt>>& getStencilCalls() {
+    return stencilCallsInOrder_;
+  }
+
+  void visit(const std::shared_ptr<StencilCallDeclStmt>& stmt) {
+    stencilCallsInOrder_.push_back(stmt);
+  }
+};
+
+/// @brief The AddBoundaryConditions class traverses the StencilDescAST to extract all the
+/// StencilCallStmts for a stencili with a given ID. This is required to properly insert boundary
+/// conditions.
+class AddBoundaryConditions : public ASTVisitorForwarding {
+  StencilInstantiation* instantiation_;
+  int StencilID_;
+
+  std::vector<std::shared_ptr<StencilCallDeclStmt>> stencilCallsToReplace_;
+
+public:
+  AddBoundaryConditions(StencilInstantiation* instantiation, int StencilID)
+      : instantiation_(instantiation), StencilID_(StencilID) {}
+
+  void visit(const std::shared_ptr<StencilCallDeclStmt>& stmt) override {
+    if(instantiation_->getStencilIDFromStmt(stmt) == StencilID_)
+      stencilCallsToReplace_.emplace_back(stmt);
+  }
+
+  std::vector<std::shared_ptr<StencilCallDeclStmt>>& getStencilCallsToReplace() {
+    return stencilCallsToReplace_;
+  }
+
+  void reset() { stencilCallsToReplace_.clear(); }
+};
+
+PassSetBoundaryCondition::PassSetBoundaryCondition() : Pass("PassSetBoundaryCondition") {}
+
+bool PassSetBoundaryCondition::run(StencilInstantiation* stencilInstantiation) {
+
+  // check if we need to run this pass
+  if(stencilInstantiation->getStencils().size() == 1) {
+    if(stencilInstantiation->getOptimizerContext()->getOptions().ReportBoundaryConditions) {
+      std::cout << "\nPASS: " << getName() << ": " << stencilInstantiation->getName() << " :";
+      std::cout << " No boundary conditions applied\n";
+    }
+    return true;
+  }
+  // returns the original ID of a variable
+  auto getOriginalID = [&](int ID) {
+    // check if the variable exists to be sure not to assert
+    auto it = stencilInstantiation->getNameToAccessIDMap().find(
+        stencilInstantiation->getOriginalNameFromAccessID(ID));
+    if(it != stencilInstantiation->getNameToAccessIDMap().end()) {
+      if(stencilInstantiation->isField(ID)) {
+        return stencilInstantiation->getAccessIDFromName(
+            stencilInstantiation->getOriginalNameFromAccessID(ID));
+      } else {
+        return (int) FieldType::FT_NotOriginal;
+      }
+    } else {
+      return (int) FieldType::FT_NotOriginal;
+    }
+
+  };
+
+  std::unordered_map<int, Extents> dirtyFields;
+  std::unordered_map<int, std::shared_ptr<BoundaryConditionDeclStmt>> allBCs;
+
+  // Fetch all the boundary conditions stored in the instantiation
+  for(const auto& bc : stencilInstantiation->getBoundaryConditions()) {
```

Comment:
all this loop can be achieved with a single line
```
std::transform(stencilInstantiation->getBoundaryConditions().begin(), stencilInstantiation->getBoundaryConditions().end(), allBCs.begin(),
                   [&](std::pair<std::string, std::shared_ptr<BoundaryConditionDeclStmt>> bcPair) -> std::pair<int, std::shared_ptr<BoundaryConditionDeclStmt>> { return stencilInstantiation->getAccessIDFromName(bcPair.first); });
```

Reply:
the key in an unordered_map is const and can therefore not be written with std::transform iirc